### PR TITLE
823 program events refactor

### LINKED ISSUE #823: Refactor program events (for querying states at a given time for reports)
Currently its only possible to get the current events.
Make it possible to query events as they have been seen at a given time.

### PR DESCRIPTION
**Pull request by clemens-msupply**
Closes #823

There is `service/src/programs/ProgramEvents.md` file that is best to be reviewed first.

### CODE REVIEW

**Comment by mark-prins** [comment]
I had some errors when testing - creating a new HIV care encounter and accessing `Physical Exam` gives

Screenshot 2022-11-22 at 1 31 52 PM

and the, if that issue is resolved:

Screenshot 2022-11-22 at 1 39 16 PM

I've put a suggested change into 2cde0dcf7

**Review by mark-prins — APPROVED** [review]
👍  looks great!

I wasn't entirely sure how to properly test - have just run through creating & updating encounters and running unit tests 🤷

**Comment by clemens-msupply** [comment]
> I had some errors when testing - creating a new HIV care encounter and accessing `Physical Exam` gives
>

Thanks! I think I had it fixed already elsewhere but think I fixed it exactly the same way so shouldn't be an issue.

File: client/packages/common/src/ui/forms/JsonForms/components/CategorizationTabLayout.tsx [hunk 1]
```diff
@@ -120,7 +120,7 @@ const UIComponent: FC<LayoutProps & AjvProps> = ({
         : categorization.elements[activeCategory]?.elements ?? [],
     schema,
     // assume the root path if not specified
-    path: path ?? '#',
+    path: path ?? '',
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply client/packages/common/src/ui/forms/JsonForms/components/CategorizationTabLayout.tsx:123]: Yes '#' didn't actually worked. Think '#' is the scope but the path is ''... :shrug:

**Pull request merged**